calyx/src/passes/group_to_seq.rs

```diff
@@ -0,0 +1,258 @@
+            let mut group_done_asmt: Option<ir::Assignment> = None;
+
+            for asmt in group.assignments.drain(..) {
+                match writes_to_cell(&asmt) {
+                    Some(cell_name) => {
+                        if OrderAnalysis::is_go_done(&asmt) {
+                            go_done_asmt = Some(asmt);
+                        } else if OrderAnalysis::is_specific_go(&asmt, &first) {
+                            first_go_asmt = Some(asmt);
+                        } else if cell_name == first {
+                            fst_asmts.push(asmt);
+                        } else if cell_name == second {
+                            snd_asmts.push(asmt);
+                        } else {
+                            unreachable!(
+                            "Does not write to one of the two \"stateful\" cells"
+                        )
+                        }
+                    }
+                    None => group_done_asmt = Some(asmt),
+                }
+            }
+
+            //Meant to hold the enable statments that will eventually
+            //form the seq that we return
+            let mut seq_vec: Vec<ir::Control> = Vec::new();
+
+            //building the first group name's prefix
+            let mut prefix = String::from("begin_split_");
+            prefix.push_str(&group_name.id);
+            let first_group = builder.add_group(prefix);
+
+            //if there is assignment in the form first.go = !first.done ? 1'd1,
+            //turn this into first.go = 1'd1.
+            match first_go_asmt {
+                None => (),
+                Some(go_asmt) => {
+                    let new_go_asmt = make_go_const(&mut builder, go_asmt);
+                    first_group.borrow_mut().assignments.push(new_go_asmt);
+                }
+            }
+
+            first_group.borrow_mut().assignments.append(&mut fst_asmts);
+
+            //spliitng a.go = b.done into group[done] = b.done and a.go = 1'd1.
+            let (group_done, cell_go) = split_go_done(
+                &mut builder,
+                go_done_asmt.unwrap_or_else(|| {
+                    unreachable!(
+                        "couldn't find a go-done assignment in {}",
+                        group_name
+                    )
+                }),
+                ir::WRC::from(&first_group),
+            );
+            first_group.borrow_mut().assignments.push(group_done);
+
+            //building second group
+            let mut prefix = String::from("end_split_");
+            prefix.push_str(&group_name.id);
+            let second_group = builder.add_group(prefix);
+            //pushing the a.go = 1'd1 that we made in split_go_done
+            second_group.borrow_mut().assignments.push(cell_go);
+            second_group.borrow_mut().assignments.append(&mut snd_asmts);
+
+            //renaming old_group[done] = ... to new_group[done] = ...
+            let new_done = rename_group_done(
```

Comment:
The right way to update a group's done condition is using `Group::done_cond_mut` instead of manually editing the assignments. The latter is unsafe and will be deprecated